full-node/sov-prover-storage-manager/src/lib.rs

```diff
@@ -0,0 +1,411 @@
+    state_snapshot_manager: Arc<RwLock<SnapshotManager>>,
+    native_snapshot_manager: Arc<RwLock<SnapshotManager>>,
+
+    phantom_mp_spec: PhantomData<S>,
+}
+
+impl<Da: DaSpec, S: MerkleProofSpec> NewProverStorageManager<Da, S> {
+    #[allow(dead_code)]
+    pub fn new(state_db: sov_schema_db::DB, native_db: sov_schema_db::DB) -> Self {
+        let snapshot_id_to_parent = Arc::new(RwLock::new(HashMap::new()));
+
+        let state_snapshot_manager = SnapshotManager::new(state_db, snapshot_id_to_parent.clone());
+        let native_snapshot_manager =
+            SnapshotManager::new(native_db, snapshot_id_to_parent.clone());
+
+        Self {
+            chain_forks: Default::default(),
+            blocks_to_parent: Default::default(),
+            latest_snapshot_id: 0,
+            block_hash_to_snapshot_id: Default::default(),
+            snapshot_id_to_parent,
+            state_snapshot_manager: Arc::new(RwLock::new(state_snapshot_manager)),
+            native_snapshot_manager: Arc::new(RwLock::new(native_snapshot_manager)),
+            phantom_mp_spec: Default::default(),
+        }
+    }
+
+    #[cfg(test)]
+    fn is_empty(&self) -> bool {
+        self.chain_forks.is_empty()
+            && self.blocks_to_parent.is_empty()
+            && self.block_hash_to_snapshot_id.is_empty()
+            && self.snapshot_id_to_parent.read().unwrap().is_empty()
+            && self.state_snapshot_manager.read().unwrap().is_empty()
+            && self.native_snapshot_manager.read().unwrap().is_empty()
+    }
+}
+
+impl<Da: DaSpec, S: MerkleProofSpec> HierarchicalStorageManager<Da>
+    for NewProverStorageManager<Da, S>
+where
+    Da::SlotHash: Hash,
+{
+    type NativeStorage = NewProverStorage<S, SnapshotManager>;
+    type NativeChangeSet = NewProverStorage<S, SnapshotManager>;
+
+    fn get_native_storage_on(
+        &mut self,
+        block_header: &Da::BlockHeader,
+    ) -> anyhow::Result<Self::NativeStorage> {
+        let current_block_hash = block_header.hash();
+        let prev_block_hash = block_header.prev_hash();
+        assert_ne!(
+            current_block_hash, prev_block_hash,
+            "Cannot provide storage for corrupt block"
+        );
+
+        let new_snapshot_id = match self.block_hash_to_snapshot_id.get(&current_block_hash) {
+            // Storage for this block has been requested before
+            Some(snapshot_id) => {
+                // TODO: Do consistency checks here?
+
+                *snapshot_id
+            }
+            // Storage requested first time
+            None => {
+                let new_snapshot_id = self.latest_snapshot_id + 1;
+                if let Some(parent_snapshot_id) =
+                    self.block_hash_to_snapshot_id.get(&prev_block_hash)
+                {
+                    let mut snapshot_id_to_parent = self.snapshot_id_to_parent.write().unwrap();
+                    snapshot_id_to_parent.insert(new_snapshot_id, *parent_snapshot_id);
+                }
```

Comment:
Should we enforcing consistency between this check and the check on line 107? In other words, is it possible for the `prev_block` to have a snapshot but not exist in the `chain_forks` map?

Reply:
Yes, there should be enforcing. It is technically possible to create snapshot id incorrect snapshot id, I plan to have test for it.

StorageManager should protect its own consistency

Reply:
I've tests all of the cases I think of, it looks this should be covered